google-cloud-clients/google-cloud-spanner/src/main/java/com/google/cloud/spanner/TransactionRunnerImpl.java

```diff
@@ -321,96 +302,98 @@ public TransactionRunner allowNestedTransaction() {
-    BackOff backoff = SpannerImpl.newBackOff();
-    final Context context = Context.current();
-    int attempt = 0;
-    // TODO: Change this to use TransactionManager.
-    while (true) {
-      checkState(
-          isValid, "TransactionRunner has been invalidated by a new operation on the session");
-      SpannerImpl.checkContext(context);
-      attempt++;
-      // TODO(user): When using streaming reads, consider using the first read to begin
-      // the txn.
-      span.addAnnotation(
-          "Starting Transaction Attempt",
-          ImmutableMap.of("Attempt", AttributeValue.longAttributeValue(attempt)));
-      txn.ensureTxn();
-
-      T result;
-      boolean shouldRollback = true;
-      try {
-        result = callable.run(txn);
-        shouldRollback = false;
-      } catch (Exception e) {
-        txnLogger.log(Level.FINE, "User-provided TransactionCallable raised exception", e);
-        if (txn.isAborted() || (e instanceof AbortedException)) {
-          span.addAnnotation(
-              "Transaction Attempt Aborted in user operation. Retrying",
-              ImmutableMap.of("Attempt", AttributeValue.longAttributeValue(attempt)));
-          shouldRollback = false;
-          backoff(context, backoff);
-          continue;
-        }
-        SpannerException toThrow;
-        if (e instanceof SpannerException) {
-          toThrow = (SpannerException) e;
-        } else {
-          toThrow = newSpannerException(ErrorCode.UNKNOWN, e.getMessage(), e);
-        }
-        span.addAnnotation(
-            "Transaction Attempt Failed in user operation",
-            ImmutableMap.<String, AttributeValue>builder()
-                .putAll(TraceUtil.getExceptionAnnotations(toThrow))
-                .put("Attempt", AttributeValue.longAttributeValue(attempt))
-                .build());
-        throw toThrow;
-      } finally {
-        if (shouldRollback) {
-          txn.rollback();
-        }
-      }
-
-      try {
-        txn.commit();
-        span.addAnnotation(
-            "Transaction Attempt Succeeded",
-            ImmutableMap.of("Attempt", AttributeValue.longAttributeValue(attempt)));
-        return result;
-      } catch (AbortedException e) {
-        txnLogger.log(Level.FINE, "Commit aborted", e);
-        span.addAnnotation(
-            "Transaction Attempt Aborted in Commit. Retrying",
-            ImmutableMap.of("Attempt", AttributeValue.longAttributeValue(attempt)));
-        backoff(context, backoff);
-      } catch (SpannerException e) {
-        span.addAnnotation(
-            "Transaction Attempt Failed in Commit",
-            ImmutableMap.<String, AttributeValue>builder()
-                .putAll(TraceUtil.getExceptionAnnotations(e))
-                .put("Attempt", AttributeValue.longAttributeValue(attempt))
-                .build());
-        throw e;
-      }
-    }
+  private <T> T runInternal(final TransactionCallable<T> txCallable) {
```

Comment:
Would you mind just briefly explaining what this change is? Hard to see from the diff. Is it mainly the return statement at the bottom?

Reply:
The web diff view is indeed not very helpful in this case...

In general: Instead of using a while(true) {...} loop for retrying the transaction, the transaction is placed in a `Callable` that can be passed to the generic retry helper. The retry helper will handle the retries if the transaction aborts.

More specifically, the change is as follows:
1. Everything inside the `while(true)` loop is moved into the `call()` method of the `Callable` retryCallable.
1. The `session.newTransaction()` call is moved from the constructor and the `backoff(...)` method to inside the `call()` method to give each attempt a fresh transaction.
1. The `continue` statement on line 354 is replaced with a `throw` statement in order to leave the method exceptionally.
1. The `Callable` is passed in to the new method `SpannerRetryHelper.runTxWithRetriesOnAborted(Callable)` that handles the retries if the transaction is aborted.